File: apps/api/real-golf-analyzer.js
#!/usr/bin/env node

/**
 * 실제 골프 비디오/사진 분석기
 * 목업 데이터 없이 진짜 사용자 미디어를 분석
 */

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const chalk = require('chalk');

class RealGolfAnalyzer {
  constructor() {
    this.pythonScript = path.join(__dirname, 'src/python/real_golf_analyzer.py');
    this.tempDir = path.join(__dirname, 'temp');
    this.ensureTempDir();
  }

  ensureTempDir() {
    if (!fs.existsSync(this.tempDir)) {
      fs.mkdirSync(this.tempDir, { recursive: true });
    }
  }

  // Base64 이미지를 실제 파일로 저장
  async saveImageFromBase64(base64Data) {
    try {
      // data:image/jpeg;base64, 제거
      const base64Clean = base64Data.replace(/^data:image\/[a-z]+;base64,/, '');
      
      // Base64를 Buffer로 변환
      const buffer = Buffer.from(base64Clean, 'base64');
      
      // 임시 파일 경로
      const filename = `temp_golf_${Date.now()}.jpg`;
      const filepath = path.join(this.tempDir, filename);
      
      // 파일 저장
      fs.writeFileSync(filepath, buffer);
      
      console.log(chalk.green(`✅ 실제 이미지 저장: ${filename} (${buffer.length} bytes)`));
      
      return filepath;
      
    } catch (error) {
      console.log(chalk.red(`❌ 이미지 저장 실패: ${error.message}`));
      return null;
    }
  }

  // 실제 Python MediaPipe로 골프 분석
  async analyzeRealGolfImage(imagePath) {
    return new Promise((resolve, reject) => {
      console.log(chalk.blue(`🔍 실제 골프 분석 시작: ${path.basename(imagePath)}`));

      const python = spawn('python3', [this.pythonScript, imagePath]);
      
      let stdout = '';
      let stderr = '';

      python.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      python.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      python.on('close', (code) => {
        // 임시 파일 삭제
        try {
          fs.unlinkSync(imagePath);
          console.log(chalk.gray(`🗑️ 임시 파일 삭제: ${path.basename(imagePath)}`));
        } catch (e) {}

        if (code === 0) {
          try {
            const result = JSON.parse(stdout);
            console.log(chalk.green(`✅ 실제 분석 완료: 포즈 감지 ${result.detected ? '성공' : '실패'}`));
            resolve(result);
          } catch (parseError) {
            console.log(chalk.red(`❌ JSON 파싱 실패: ${parseError.message}`));
            console.log('Python 출력:', stdout);
            resolve({ success: false, error: 'JSON parse error' });
          }
        } else {
          console.log(chalk.red(`❌ Python 실행 실패 (코드 ${code})`));
          console.log('에러:', stderr);
          resolve({ success: false, error: 'Python execution failed' });
        }
      });
    });
  }

  // 메인 분석 함수
  async analyze(mediaData) {
    try {
      console.log(chalk.blue.bold('🏌️ 실제 골프 분석 시작'));
      
      // 1. Base64 이미지를 실제 파일로 저장
      const imagePath = await this.saveImageFromBase64(mediaData);
      if (!imagePath) {
        return {
          success: false,
          error: '이미지 저장 실패',
          message: '유효한 이미지 데이터가 아닙니다'
        };
      }

      // 2. 실제 Python MediaPipe로 분석
      const pythonResult = await this.analyzeRealGolfImage(imagePath);
      
      if (!pythonResult.success) {
        return {
          success: false,
          error: pythonResult.error,
          message: '골프 분석에 실패했습니다'
        };
      }

      // 3. 실제 분석 결과 처리
      return this.processRealAnalysisResult(pythonResult);

    } catch (error) {
      console.log(chalk.red(`❌ 분석 오류: ${error.message}`));
      return {
        success: false,
        error: error.message,
        message: '예상치 못한 오류가 발생했습니다'
      };
    }
  }

  // 실제 분석 결과 처리
  processRealAnalysisResult(pythonResult) {
    console.log(chalk.cyan('📊 실제 분석 결과 처리 중...'));

    if (!pythonResult.detected) {
      return {
        success: false,
        message: '골프 자세를 감지할 수 없습니다',
        suggestions: [
          '전신이 보이도록 촬영해주세요',
          '골프 어드레스 자세를 취해주세요',
          '조명이 충분한 곳에서 촬영해주세요'
        ]
      };
    }

    // 실제 랜드마크 데이터 기반 분석
    const landmarks = pythonResult.landmarks || [];
    const keyPoints = pythonResult.key_points || {};

    const analysis = {
      success: true,
      data: {
        // 실제 계산된 점수
        score: pythonResult.score || this.calculateRealScore(keyPoints),
        
        // 실제 감지된 자세 분석
        posture: {
          detected: pythonResult.detected,
          confidence: pythonResult.confidence || 0,
          landmarks_count: landmarks.length,
          key_points: keyPoints
        },

        // 실제 각도 측정값
        angles: pythonResult.angles || this.calculateRealAngles(keyPoints),
        
        // 실제 피드백 (목업 아님)
        feedback: this.generateRealFeedback(keyPoints, pythonResult.angles),
        
        // 실제 개선사항
        improvements: this.generateRealImprovements(keyPoints, pythonResult.angles),

        // 처리 정보
        processing: {
          method: 'Real MediaPipe Analysis',
          landmarks_detected: landmarks.length,
          confidence: pythonResult.confidence || 0,
          analysis_type: 'Actual Golf Posture Analysis',
          data_source: 'User Uploaded Image - Real Analysis'
        }
      }
    };

    console.log(chalk.green(`✅ 실제 분석 완료: ${landmarks.length}개 랜드마크, 점수 ${analysis.data.score}`));
    return analysis;
  }

  // 실제 점수 계산
  calculateRealScore(keyPoints) {
    let score = 50; // 기본 점수

    // 실제 키포인트 기반 점수 계산
    if (keyPoints.head && keyPoints.shoulders) {
      score += 10; // 머리와 어깨 감지됨
    }
    
    if (keyPoints.hips && keyPoints.knees) {
      score += 10; // 힙과 무릎 감지됨
    }
    
    if (keyPoints.hands && keyPoints.elbows) {
      score += 15; // 팔 자세 감지됨
    }
    
    if (keyPoints.feet) {
      score += 10; // 발 위치 감지됨
    }

    return Math.min(95, score);
  }

  // 실제 각도 계산
  calculateRealAngles(keyPoints) {
    // 실제 3D 좌표 기반 각도 계산
    return {
      spine_angle: keyPoints.spine_angle || 0,
      shoulder_tilt: keyPoints.shoulder_tilt || 0,
      hip_angle: keyPoints.hip_angle || 0,
      knee_flex: keyPoints.knee_flex || 0
    };
  }

  // 실제 피드백 생성
  generateRealFeedback(keyPoints, angles) {
    const feedback = [];

    if (!keyPoints.head) {
      feedback.push('머리 위치를 확인할 수 없습니다');
    }
    
    if (!keyPoints.shoulders) {
      feedback.push('어깨 라인이 명확하지 않습니다');
    }
    
    if (angles && angles.spine_angle > 35) {
      feedback.push('척추가 너무 많이 기울어져 있습니다');
    }
    
    if (angles && angles.spine_angle < 15) {
      feedback.push('척추를 조금 더 기울여주세요');
    }

    return feedback.length > 0 ? feedback : ['자세가 전반적으로 좋아 보입니다'];
  }

  // 실제 개선사항 생성
  generateRealImprovements(keyPoints, angles) {
    const improvements = [];

    if (!keyPoints.feet) {
      improvements.push('발 전체가 보이도록 촬영해주세요');
    }
    
    if (keyPoints.confidence < 0.7) {
      improvements.push('더 밝은 조명에서 촬영해보세요');
    }
    
    if (angles && angles.shoulder_tilt > 10) {
      improvements.push('어깨를 수평으로 맞춰보세요');
    }

    return improvements.length > 0 ? improvements : ['현재 자세를 유지하시기 바랍니다'];
  }
}

module.exports = RealGolfAnalyzer;

File: apps/api/src/python/real_golf_analyzer.py
#!/usr/bin/env python3
"""
실제 골프 이미지/비디오 분석기
목업 데이터 없이 실제 MediaPipe로 분석
"""

import cv2
import numpy as np
import json
import sys
import os
import mediapipe as mp
import math
from pathlib import Path

class RealGolfAnalyzer:
    def __init__(self):
        # MediaPipe 초기화
        self.mp_pose = mp.solutions.pose
        self.mp_drawing = mp.solutions.drawing_utils
        
        # 실제 골프 분석용 설정 (높은 정확도)
        self.pose = self.mp_pose.Pose(
            static_image_mode=True,
            model_complexity=2,  # 최고 복잡도
            min_detection_confidence=0.6,  # 골프 자세 감지용
            min_tracking_confidence=0.5,
            enable_segmentation=True,
            smooth_segmentation=True
        )
        
        # 골프 관련 핵심 랜드마크 (MediaPipe 33개 포인트 중)
        self.golf_keypoints = {
            'nose': 0,
            'left_shoulder': 11, 'right_shoulder': 12,
            'left_elbow': 13, 'right_elbow': 14,
            'left_wrist': 15, 'right_wrist': 16,
            'left_hip': 23, 'right_hip': 24,
            'left_knee': 25, 'right_knee': 26,
            'left_ankle': 27, 'right_ankle': 28
        }
    
    def load_image(self, image_path):
        """실제 이미지 파일 로드"""
        try:
            if not os.path.exists(image_path):
                return None, f"파일이 존재하지 않습니다: {image_path}"
            
            # OpenCV로 이미지 읽기
            image = cv2.imread(image_path)
            if image is None:
                return None, "이미지를 읽을 수 없습니다"
            
            # RGB로 변환 (MediaPipe는 RGB 사용)
            image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
            
            return image_rgb, None
            
        except Exception as e:
            return None, f"이미지 로드 오류: {str(e)}"
    
    def analyze_pose(self, image):
        """실제 포즈 분석 (목업 아님)"""
        try:
            # MediaPipe로 실제 포즈 감지
            results = self.pose.process(image)
            
            if not results.pose_landmarks:
                return {
                    'detected': False,
                    'confidence': 0,
                    'landmarks': [],
                    'error': 'No pose detected in image'
                }
            
            # 실제 랜드마크 데이터 추출
            landmarks = []
            key_points = {}
            
            for idx, landmark in enumerate(results.pose_landmarks.landmark):
                # 실제 3D 좌표 (x, y, z, visibility)
                landmarks.append({
                    'id': idx,
                    'x': landmark.x,
                    'y': landmark.y,
                    'z': landmark.z,
                    'visibility': landmark.visibility
                })
                
                # 골프 관련 키포인트 추출
                for name, point_id in self.golf_keypoints.items():
                    if idx == point_id:
                        key_points[name] = {
                            'x': landmark.x,
                            'y': landmark.y, 
                            'z': landmark.z,
                            'visibility': landmark.visibility
                        }
            
            # 평균 confidence 계산
            avg_confidence = np.mean([lm['visibility'] for lm in landmarks])
            
            return {
                'detected': True,
                'confidence': float(avg_confidence),
                'landmarks': landmarks,
                'key_points': key_points,
                'landmark_count': len(landmarks)
            }
            
        except Exception as e:
            return {
                'detected': False,
                'confidence': 0,
                'error': f'Pose analysis failed: {str(e)}'
            }
    
    def calculate_real_angles(self, key_points):
        """실제 골프 각도 계산 (3D 좌표 기반)"""
        try:
            angles = {}
            
            # 척추 각도 (목-허리 라인)
            if 'nose' in key_points and 'left_hip' in key_points and 'right_hip' in key_points:
                nose = key_points['nose']
                left_hip = key_points['left_hip']
                right_hip = key_points['right_hip']
                
                # 힙 중점
                hip_center = {
                    'x': (left_hip['x'] + right_hip['x']) / 2,
                    'y': (left_hip['y'] + right_hip['y']) / 2
                }
                
                # 척추 각도 계산
                spine_angle = math.atan2(
                    abs(nose['x'] - hip_center['x']),
                    abs(nose['y'] - hip_center['y'])
                ) * 180 / math.pi
                
                angles['spine_angle'] = spine_angle
            
            # 어깨 기울기
            if 'left_shoulder' in key_points and 'right_shoulder' in key_points:
                left_shoulder = key_points['left_shoulder']
                right_shoulder = key_points['right_shoulder']
                
                shoulder_tilt = math.atan2(
                    right_shoulder['y'] - left_shoulder['y'],
                    right_shoulder['x'] - left_shoulder['x']
                ) * 180 / math.pi
                
                angles['shoulder_tilt'] = abs(shoulder_tilt)
            
            # 무릎 굽힘 (간단 근사)
            if 'left_hip' in key_points and 'left_knee' in key_points and 'left_ankle' in key_points:
                hip = key_points['left_hip']
                knee = key_points['left_knee']
                ankle = key_points['left_ankle']
                
                # 무릎 각도 계산
                vec1 = np.array([hip['x'] - knee['x'], hip['y'] - knee['y']])
                vec2 = np.array([ankle['x'] - knee['x'], ankle['y'] - knee['y']])
                
                cos_angle = np.dot(vec1, vec2) / (np.linalg.norm(vec1) * np.linalg.norm(vec2))
                cos_angle = np.clip(cos_angle, -1.0, 1.0)  # 수치 안정성
                knee_angle = math.acos(cos_angle) * 180 / math.pi
                
                angles['knee_flex'] = 180 - knee_angle  # 굽힘 각도
            
            return angles
            
        except Exception as e:
            print(f"각도 계산 오류: {e}", file=sys.stderr)
            return {}
    
    def evaluate_golf_posture(self, key_points, angles):
        """실제 골프 자세 평가"""
        score = 50  # 기본 점수
        feedback = []
        improvements = []
        
        try:
            # 포즈 감지 기본 점수
            if len(key_points) >= 8:  # 주요 키포인트 감지
                score += 20
                feedback.append("주요 신체 부위가 잘 감지되었습니다")
            else:
                improvements.append("전신이 더 잘 보이도록 촬영해주세요")
            
            # 척추 각도 평가
            if 'spine_angle' in angles:
                spine = angles['spine_angle']
                if 15 <= spine <= 35:  # 이상적인 범위
                    score += 15
                    feedback.append(f"척추 각도가 좋습니다 ({spine:.1f}°)")
                elif spine < 15:
                    score += 5
                    improvements.append("척추를 조금 더 앞으로 기울여보세요")
                elif spine > 35:
                    score += 5
                    improvements.append("척추를 조금 더 세워보세요")
            
            # 어깨 기울기 평가
            if 'shoulder_tilt' in angles:
                tilt = angles['shoulder_tilt']
                if tilt <= 10:  # 수평에 가까움
                    score += 10
                    feedback.append("어깨 라인이 좋습니다")
                else:
                    improvements.append("어깨를 더 수평으로 맞춰보세요")
            
            # 무릎 굽힘 평가
            if 'knee_flex' in angles:
                flex = angles['knee_flex']
                if 10 <= flex <= 30:  # 적절한 굽힘
                    score += 10
                    feedback.append("무릎 굽힘이 적절합니다")
                else:
                    improvements.append("무릎을 자연스럽게 굽혀보세요")
            
            # 양발 감지 확인
            if 'left_ankle' in key_points and 'right_ankle' in key_points:
                score += 5
                feedback.append("발 위치가 잘 감지되었습니다")
            else:
                improvements.append("발 전체가 보이도록 촬영해주세요")
            
            # 최종 점수 범위 조정
            score = max(30, min(95, score))
            
            if len(feedback) == 0:
                feedback = ["자세 분석이 완료되었습니다"]
            
            if len(improvements) == 0:
                improvements = ["현재 자세를 유지하시기 바랍니다"]
            
            return score, feedback, improvements
            
        except Exception as e:
            print(f"자세 평가 오류: {e}", file=sys.stderr)
            return 50, ["자세 분석 중 오류가 발생했습니다"], ["다시 촬영해보시기 바랍니다"]
    
    def analyze_image(self, image_path):
        """메인 분석 함수"""
        try:
            # 1. 실제 이미지 로드
            image, error = self.load_image(image_path)
            if image is None:
                return {
                    'success': False,
                    'error': error,
                    'detected': False
                }
            
            print(f"이미지 로드 성공: {image.shape}", file=sys.stderr)
            
            # 2. 실제 포즈 분석
            pose_result = self.analyze_pose(image)
            
            if not pose_result['detected']:
                return {
                    'success': False,
                    'detected': False,
                    'error': pose_result.get('error', 'Pose not detected'),
                    'message': '골프 자세를 감지할 수 없습니다'
                }
            
            print(f"포즈 감지 성공: {pose_result['landmark_count']}개 랜드마크", file=sys.stderr)
            
            # 3. 실제 각도 계산
            angles = self.calculate_real_angles(pose_result['key_points'])
            print(f"각도 계산 완료: {len(angles)}개", file=sys.stderr)
            
            # 4. 실제 자세 평가
            score, feedback, improvements = self.evaluate_golf_posture(
                pose_result['key_points'], 
                angles
            )
            
            # 5. 최종 결과 반환 (목업 데이터 없음)
            result = {
                'success': True,
                'detected': True,
                'confidence': pose_result['confidence'],
                'landmarks': pose_result['landmarks'],
                'key_points': pose_result['key_points'],
                'angles': angles,
                'score': score,
                'feedback': feedback,
                'improvements': improvements,
                'analysis_info': {
                    'method': 'Real MediaPipe Analysis',
                    'landmarks_count': pose_result['landmark_count'],
                    'image_size': f"{image.shape[1]}x{image.shape[0]}",
                    'confidence': pose_result['confidence']
                }
            }
            
            print(f"분석 완료: 점수 {score}, 신뢰도 {pose_result['confidence']:.2f}", file=sys.stderr)
            return result
            
        except Exception as e:
            error_msg = f"분석 중 오류 발생: {str(e)}"
            print(error_msg, file=sys.stderr)
            return {
                'success': False,
                'detected': False,
                'error': error_msg
            }

def main():
    if len(sys.argv) < 2:
        print(json.dumps({
            'success': False,
            'error': 'Image path required'
        }))
        return
    
    image_path = sys.argv[1]
    
    # 실제 분석 실행
    analyzer = RealGolfAnalyzer()
    result = analyzer.analyze_image(image_path)
    
    # JSON 결과 출력
    print(json.dumps(result, indent=2))

if __name__ == "__main__":
    main()